### PR TITLE
feat(statement mutator): Add mutator that removes statements

### PR DESCRIPTION
**Pull request by ivancea**
issue: #1470

### CODE REVIEW

**Comment by ivancea** [comment]
Still working on MutantOrchestratorTests tests that didn't pass after this mutator

File: src/Stryker.Core/Stryker.Core/Mutators/Mutator.cs [hunk 1]
```diff
@@ -27,6 +27,8 @@ public enum Mutator
         Linq,
         [Description("String literals")]
         String,
+        [Description("Statements")]
```
Comment (ivancea) [inline src/Stryker.Core/Stryker.Core/Mutators/Mutator.cs:30]: Not sure about the order of these elements. Same for CsharpMutantOrchestrator

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            new object[] {
+                SyntaxFactory.IfStatement(
+                    SyntaxFactory.LiteralExpression(SyntaxKind.TrueLiteralExpression),
+                    SyntaxFactory.EmptyStatement()
+                )
+            },
+        };
+
+        [Theory]
+        [MemberData(nameof(MutableStatements))]
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:43]: We prefer to use InlineData instead of MemberData. And since the `member` is not used anywhere else I'd like to see InlineData here as well

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+        [Fact]
+        private void ShouldNotMutate()
+        {
+            SyntaxTree tree = CSharpSyntaxTree.ParseText($@"
+namespace Test
+{{
+    class Program
+    {{
+        delegate void TestDelegate();
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:64]: Why is this here?

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+        static int Method()
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:66]: I'm missing the empty statement?

**Review by richardwerkman — CHANGES_REQUESTED** [review]
Thanks for creating this mutator! This will be a nice addition to stryker 😄 

Ps. You will have to update the integration tests as well, since this will introduce a lot of new mutations. We should probably set the mutation level to `Complete` in the integration test to make sure we also test these mutations.

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 2 above)
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:43]: I used MemberData because the statements aren't constants, and I cannot construct them inside InlineData (unless I'm missing something)

**Comment by ivancea** [comment]
I'll have to rethink this mutator completely. I'm finding compilation problems that aren't being rollbacked and are throwing errors:

1. Any variable assignation (and any statement having an asignation inside) may result in an error if they are assigning an unassigned variable. Example:
```Cs
string v;
if (..) {
    v = "";
} else {
    v = " ";
}
```

2. The last statement inside a switch case will throw an error if removed (return/throw/break/goto)

I'll switch from the "forbidden syntaxes" approach to a "allowed syntaxes" one, and document avery case I find. The most important problem now is the assignment thing, as it isn't trivial to find which assignments (or blocks containing them) are removable. As in devMode they are resulting in a critical error, I'm assumming it's not acceptable. I'm converting this PR to draft

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace Stryker.Core.UnitTest.Mutators
+{
+    public class StatementMutatorTests
+    {
+        [Fact]
+        public void ShouldBeMutationlevelStandard()
+        {
+            var target = new StatementMutator();
+            target.MutationLevel.ShouldBe(MutationLevel.Standard);
```
[the remark this answers is not in the document]
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:18]: After the rework, it added about ~25% extra mutators to a project I tested (NFluent, ~2000 -> ~2500).
So, as you like. I would say it's ok in standard. Method calls may produce untested side effects, as well as breaks returns..., but there are some "repeated" mutations (x++; -> x--; & ;) that may be considered redundant, so maybe advanced?

File: src/Stryker.Core/Stryker.Core/Mutators/StatementMutator.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+using Stryker.Core.Mutants;
+using System.Linq;
+
+namespace Stryker.Core.Mutators
+{
+    public class StatementMutator: MutatorBase<StatementSyntax>, IMutator
+    {
+        public override MutationLevel MutationLevel => MutationLevel.Standard;
+
+        private static readonly HashSet<SyntaxKind> ForbiddenSyntaxes = new HashSet<SyntaxKind>() {
```
[the remark this answers is not in the document]
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core/Mutators/StatementMutator.cs]: 100% reworked to just allow some syntaxes

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 3 above)
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs]: Removed. used it to check lambdas and anonymous methods, but it isn't necessary now

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 4 above)
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:64]: I'm not sure what you mean. Te test it isn't mutated?
After the rework, with a whitelist of mutable statements, I'm not sure if I should add every non-whitelisted statement to the test. What do you think? Every other StatementSyntax is commented in the StatementMutator whitelist

**Comment by ivancea** [comment]
Reopenign after reworking it :)

**Comment by ivancea** [comment]
Not sure about why are the Windows integration tests failing

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Mobrockers) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:18]: With the rewrite of this mutator we think standard is fine for now.

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 2 above)
Comment (Mobrockers) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs:43]: If you look at other tests you'll see that we place the syntax as string in InlineData and parse it to syntax in the unit test. This is more readable because you can clearly see what syntax we're testing instead of having to build up a mental model from a SyntaxTree.

**Comment by Mobrockers** [comment]
> Not sure about why are the Windows integration tests failing

It is not the integration test that is failing, it is a unit test:

https://dev.azure.com/stryker-mutator/Stryker/_build/results?buildId=6939&view=ms.vss-test-web.build-test-results-tab&runId=1067866&resultId=100404&paneView=debug

```
Shouldly.ShouldAssertException : isSame
should be
True
but was
False

Additional Info:
AST's are not equivalent. Line[1]
actual:(Stryker6Gp9tRNthZmz0ED.MutantControl.IsActive(ID)?1-8:1+8)
expect:(Stryker6Gp9tRNthZmz0ED.MutantControl.IsActive(16)?1-8:1+8)
Actual(full):
(Stryker6Gp9tRNthZmz0ED.MutantControl.IsActive(ID)?1-8:1+8)
, expected:
(Stryker6Gp9tRNthZmz0ED.MutantControl.IsActive(16)?1-8:1+8)
```

```
at Stryker.Core.UnitTest.AssertionExtensions.ShouldBeSemantically(SyntaxNode actual, SyntaxNode expected) in D:\a\1\s\src\Stryker.Core\Stryker.Core.UnitTest\AssertExtensions.cs:line 52
   at Stryker.Core.UnitTest.AssertionExtensions.ShouldBeSemantically(String actual, String expected) in D:\a\1\s\src\Stryker.Core\Stryker.Core.UnitTest\AssertExtensions.cs:line 22
   at Stryker.Core.UnitTest.Mutants.MutantPlacerTests.MutantPlacer_ShouldPlaceWithConditionalExpression(Int32 id) in D:\a\1\s\src\Stryker.Core\Stryker.Core.UnitTest\Mutants\MutantPlacerTests.cs:line 61
```

File: src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs (hunk 2 above)
Comment (ivancea) [reply src/Stryker.Core/Stryker.Core.UnitTest/Mutators/StatementMutatorTests.cs]: Ok, changed

**Comment by ivancea** [comment]
The unit test that were failing in the pipeline (not in my local) isn't failing now; I don't know what happened 🙃

File: src/Stryker.Core/Stryker.Core/Mutators/StatementMutator.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    public class StatementMutator: MutatorBase<StatementSyntax>, IMutator
+    {
+        public override MutationLevel MutationLevel => MutationLevel.Standard;
+
+        private static readonly HashSet<SyntaxKind> AllowedSyntaxes = new HashSet<SyntaxKind>()
+        {
+            // SyntaxKind.EmptyStatement, // useless mutation
+
+            // SyntaxKind.Block, // unitary mutation should prevail over blocks
+            // SyntaxKind.CheckedStatement, // not unitary
```
Comment (Mobrockers) [inline src/Stryker.Core/Stryker.Core/Mutators/StatementMutator.cs:18]: Could you clarify what the comment not unitary means on these types of statements

Comment (ivancea) [reply src/Stryker.Core/Stryker.Core/Mutators/StatementMutator.cs:18]: As "not unitary" I meant every statement that contains more statements inside (usually, containing blocks)
